lib/CountdownBarrier.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+namespace Amp;
+
+use Error;
+
+/**
+ * A countdown barrier returns a promise that is resolved when it was signaled n times.
+ *
+ * **Example**
+ *
+ * ```php
+ * $countdownBarrier = new \Amp\CountdownBarrier(2);
+ * $countdownBarrier->signal();
+ * $countdownBarrier->signal(); // promise is now resolved
+ * ```
+ */
+final class CountdownBarrier
+{
+    /** @var int */
+    private $initialCount;
+    /** @var int */
+    private $currentCount;
+    /** @var Deferred */
+    private $deferred;
+
+    public function __construct(int $initialCount)
+    {
+        if ($initialCount < 1) {
+            throw new Error('Counter must be positive');
+        }
+
+        $this->initialCount = $initialCount;
+        $this->currentCount = $initialCount;
+        $this->deferred = new Deferred();
+    }
+
+    public function signal(int $signalCount = 1): bool
+    {
+        if ($signalCount < 1) {
+            throw new Error('Signal count must be greater or equals 1');
+        }
+
+        if (0 === $this->currentCount) {
+            throw new Error('CountdownBarrier already resolved');
+        }
+
+        $this->currentCount -= $signalCount;
+
+        if (0 === $this->currentCount) {
+            $this->deferred->resolve(true);
+
+            return true;
+        }
```

Comment:
This should check for `$this->currentCount < 0`  and throw an error then, too.

Reply:
Throw or just make it 0?